### PR TITLE
Explain planning of text indexes

### PR DESCRIPTION
**Pull request by hedengran**
Depends on https://github.com/neo-technology/neo4j/pull/12568

### CODE REVIEW

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala [hunk 1]
```diff
@@ -348,6 +365,38 @@ class SchemaIndexTest extends DocumentingTestBase with QueryStatisticsTestSuppor
     )
   }
 
+  @Test def use_relationship_text_index() {
+    profileQuery(
+      title = "Relationship TEXT index example",
+      text = "In this example, a `KNOWS(lastMetLocation)` relationship TEXT index is available. ",
```
Comment (Hunterness) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala]:
```suggestion
      text = "In this example, a `KNOWS(lastMetLocation)` relationship `TEXT` index is available. ",
```

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala [hunk 2]
```diff
@@ -317,10 +320,10 @@ class SchemaIndexTest extends DocumentingTestBase with QueryStatisticsTestSuppor
     )
   }
 
-  @Test def use_index() {
+  @Test def use_btree_index() {
     profileQuery(
-      title = "Node index example",
-      text = "In the example below, the query uses a `Person(firstname)` node index, if it exists. ",
+      title = "Node BTREE index example",
+      text = "In the example below, a `Person(firstname)` `BTREE` node index is available.",
```
Comment (Hunterness) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala]: Here you have `BTREE node index`, but then you have `relationship BTREE index` in the relationship example (same for TEXT), maybe be consistent?

File: cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc [hunk 3]
```diff
@@ -15,10 +15,83 @@ Therefore, node and relationship indexes are used interchangeably in this sectio
...
+There are different types of indexes available in Neo4j.
+Not all indexes are compatible with the same property predicates.
 
-include::ql/administration/indexes/relationship-index-example.asciidoc[leveloffset=2]
+In the most common case, indexes are used for `MATCH` and `OPTIONAL MATCH` clauses that combine a label predicate with a property predicate.
+Therefore, it is important to know the kind of predicates that can be solved by the different indexes.
+
+=== BTREE indexes
+
+`BTREE` indexes support all types of predicates:
```
Comment (Hunterness) [inline cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc:27]: Should we mention that BTREE allows composite indexes while text doesn't?

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala [hunk 4]
```diff
@@ -331,10 +334,24 @@ class SchemaIndexTest extends DocumentingTestBase with QueryStatisticsTestSuppor
     )
   }
 
-  @Test def use_relationship_index() {
+  @Test def use_text_index() {
```
Comment (Hunterness) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala:382]: do we want to update to/add text index tests to the ref card as well?

File: cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc (hunk 3 above)
Comment (PetrJanouch) [reply cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc:27]: I have been thinking about it, too.
It is a question if this is the place. We don't allow creating a composite text index, so a user will definitely not be wondering "why did the planner not pick up my composite text index?"

**Review by PetrJanouch — APPROVED** [review]
Looks good to me

File: cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc (hunk 3 above)
Comment (Hunterness) [reply cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc:27]: true

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala (hunk 2 above)
Comment (hedengran) [reply cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala]: Good point, thanks!

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala (hunk 4 above)
Comment (hedengran) [reply cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala:382]: Good point, thanks!

File: cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc [hunk 5]
```diff
@@ -15,10 +15,78 @@ Therefore, node and relationship indexes are used interchangeably in this sectio
 
 Please refer to <<administration-indexes-search-performance>> for instructions on how to create and maintain the indexes themselves.
 
+== Index types and predicate compatibility
 
-include::ql/administration/indexes/node-index-example.asciidoc[leveloffset=2]
+There are different types of indexes available in Neo4j.
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc]:
```suggestion
There are different types of indexes available in Neo4j but they are not all compatible with the same property predicates.
```

File: cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc [hunk 6]
```diff
@@ -15,10 +15,78 @@ Therefore, node and relationship indexes are used interchangeably in this sectio
...
+Not all indexes are compatible with the same property predicates.
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc]:
```suggestion

```

File: cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc [hunk 7]
```diff
@@ -15,10 +15,78 @@ Therefore, node and relationship indexes are used interchangeably in this sectio
...
 
-include::ql/administration/indexes/relationship-index-example.asciidoc[leveloffset=2]
+In the most common case, indexes are used for `MATCH` and `OPTIONAL MATCH` clauses that combine a label predicate with a property predicate.
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc]:
```suggestion
Indexes are commonly used for `MATCH` and `OPTIONAL MATCH` clauses that combine a label predicate with a property predicate.
```

File: cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc [hunk 8]
```diff
@@ -15,10 +15,78 @@ Therefore, node and relationship indexes are used interchangeably in this sectio
...
+Therefore, it is important to know the kind of predicates that can be solved by the different indexes.
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc]:
```suggestion
Therefore, it is important to know what kind of predicates that can be solved by the different indexes.
```

File: cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc [hunk 9]
```diff
@@ -15,10 +15,78 @@ Therefore, node and relationship indexes are used interchangeably in this sectio
...
+|range search| `n.prop > value`
+|prefix search| `STARTS WITH`
+|suffix search| `ENDS WITH`
+|substring search| `CONTAINS`
+|===
+
+=== TEXT indexes
+
+`TEXT` indexes only work for predicates operating on strings.
+That means that `TEXT` indexes will only be used when it is known that the predicate will evaluate to `null` for all non-string values.
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc]:
```suggestion
That means that `TEXT` indexes are only used when it is known that the predicate evaluates to `null` for all non-string values.
```

File: cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc [hunk 10]
```diff
@@ -15,10 +15,78 @@ Therefore, node and relationship indexes are used interchangeably in this sectio
...
+
+Predicates that only operate on strings are always solvable by a `TEXT` index:
+
+* `STARTS WITH`
+* `ENDS WITH`
+* `CONTAINS`
+
+Other predicates, however, will only be used when it is known that the property is compared to a string:
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc]:
```suggestion
However, other predicates are only used when it is known that the property is compared to a string:
```

File: cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc [hunk 11]
```diff
@@ -15,10 +15,78 @@ Therefore, node and relationship indexes are used interchangeably in this sectio
...
+
+* `n.prop = "string"`
+* `n.prop IN ["a", "b", "c"]`
+* `n.prop > "string"`
+
+That means that a `TEXT` index will not be able to solve e.g. `a.prop = b.prop`.
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc]:
```suggestion
This means that a `TEXT` index is not able to solve e.g. `a.prop = b.prop`.
```

File: cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc [hunk 12]
```diff
@@ -15,10 +15,78 @@ Therefore, node and relationship indexes are used interchangeably in this sectio
...
+|list membership check | `n.prop IN ["a", "b", "c"]`
+|range search| `n.prop > "string"`
+|prefix search| `STARTS WITH`
+|suffix search| `ENDS WITH`
+|substring search| `CONTAINS`
+|===
+
+== Index preference
+
+When multiple indexes are available and able to solve a predicate there is an ordering defined that decide on which index to use.
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc]:
```suggestion
When multiple indexes are available and able to solve a predicate, there is an order defined that decides which index to use.
```

File: cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc [hunk 13]
```diff
@@ -15,10 +15,78 @@ Therefore, node and relationship indexes are used interchangeably in this sectio
...
+It is defined such that:
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc]:
```suggestion
It is defined as such:
```

File: cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc [hunk 14]
```diff
@@ -15,10 +15,78 @@ Therefore, node and relationship indexes are used interchangeably in this sectio
...
+
+* `TEXT` indexes will be preferred over `BTREE` indexes.
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc]:
```suggestion
* `TEXT` indexes are preferred over `BTREE` indexes.
```

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala [hunk 15]
```diff
@@ -386,6 +410,38 @@ class SchemaIndexTest extends DocumentingTestBase with QueryStatisticsTestSuppor
...
+          assertEquals(1, p.size)
+
+          checkPlanDescription(p)("DirectedRelationshipIndexSeek")
+      }
+    )
+  }
+
+  @Test def use_text_over_btree_index() {
+    profileQuery(
+      title = "When multiple index types are available",
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala]:
```suggestion
      title = "Multiple available index types",
```

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala [hunk 16]
```diff
@@ -386,6 +410,38 @@ class SchemaIndexTest extends DocumentingTestBase with QueryStatisticsTestSuppor
...
+      text = "In the example below, both a `Person(middlename)` node `TEXT` index and a `Person(middlename)` node `BTREE` index are available. " +
+        "The `TEXT` node index will be chosen.",
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SchemaIndexTest.scala]:
```suggestion
        "The `TEXT` node index is chosen.",
```

File: cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/IndexTest.scala [hunk 17]
```diff
@@ -185,9 +185,32 @@ MATCH (n:Person) WHERE n.name = $$value
...
+###assertion=match
+//
+
+MATCH (n:Person) WHERE n.name = "Alice"
+
+RETURN n
+###
+
+An TEXT index can be automatically used for the equality comparison when comparing to a string.
+Note that for example `toLower(n.name) = "string"` will not use an index.
```
Comment (AlexicaWright) [inline cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/IndexTest.scala]:
```suggestion
Note that for example `toLower(n.name) = "string"` does not use an index.
```

File: cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/IndexTest.scala [hunk 18]
```diff
@@ -185,9 +185,32 @@ MATCH (n:Person) WHERE n.name = $$value
...
+//
+
+MATCH (n:Person)
+WHERE n.name < "Bob"
+
+RETURN n
+###
+
+An index can automatically be used for range predicates.
+Note that a TEXT index will only be used if the predicate compares the property with a string.
```
Comment (AlexicaWright) [inline cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/IndexTest.scala]:
```suggestion
Note that a TEXT index is only used if the predicate compares the property with a string.
```

**Review by AlexicaWright — COMMENTED** [review]
Some editorial comments to consider.

File: cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/IndexTest.scala [hunk 19]
```diff
@@ -185,9 +185,32 @@ MATCH (n:Person) WHERE n.name = $$value
 RETURN n
 ###
 
-An index can be automatically used for the equality comparison.
+An BTREE index can be automatically used for the equality comparison.
```
Comment (tobias-johansson) [inline cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/IndexTest.scala:188]:
```suggestion
A BTREE index can be automatically used for the equality comparison.
```

File: cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/IndexTest.scala [hunk 20]
```diff
@@ -185,9 +185,32 @@ MATCH (n:Person) WHERE n.name = $$value
...
 
+###assertion=match
+//
+
+MATCH (n:Person) WHERE n.name = "Alice"
+
+RETURN n
+###
+
+An TEXT index can be automatically used for the equality comparison when comparing to a string.
```
Comment (tobias-johansson) [inline cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/IndexTest.scala:199]:
```suggestion
A TEXT index can be automatically used for the equality comparison when comparing to a string.
```

Comment (tobias-johansson) [reply cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/IndexTest.scala:199]: Should index types have code formatting?
```suggestion
A `TEXT` index can be automatically used for the equality comparison when comparing to a string.
```

File: cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/IndexTest.scala (hunk 19 above)
Comment (tobias-johansson) [reply cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/IndexTest.scala:188]: Should index types have code formatting?
```suggestion
A `BTREE` index can be automatically used for the equality comparison.
```

File: cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/IndexTest.scala [hunk 21]
```diff
@@ -199,6 +222,17 @@ RETURN n
...
+###assertion=match parameters=aname
+//
+
+MATCH (n:Person)
+WHERE n.name IN ['Bob', 'Alice']
+
+RETURN n
+###
+
+An TEXT index can automatically be used for the `IN` list checks when all elements in the list are strings.
```
Comment (tobias-johansson) [inline cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/IndexTest.scala:234]:
```suggestion
A TEXT index can automatically be used for the `IN` list checks when all elements in the list are strings.
```

**Review by tobias-johansson — COMMENTED** [review]
Looks great! Found a few minor spelling things

**Review by AlexicaWright — APPROVED** [review]
👍 LGTM now! Thank you @hedengran !

File: cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc [hunk 22]
```diff
@@ -15,10 +15,77 @@ Therefore, node and relationship indexes are used interchangeably in this sectio
...
+|suffix search| `ENDS WITH`
+|substring search| `CONTAINS`
+|===
+
+== Index preference
+
+When multiple indexes are available and able to solve a predicate, there is an order defined that decides which index to use.
+It is defined as such:
+
+* `TEXT` indexes are preferred over `BTREE` indexes.
```
Comment (burqen) [inline cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc:78]: TEXT should only be preferred over BTREE for CONTAINS and ENDS WITH. In all other cases BTREE should be more efficient. This should be reflected by `IndexCapability#getCostMultiplier( IndexQueryType... queryTypes )`. Is that not utilised by the planner? Ping @hedengran

**Pull request merged**